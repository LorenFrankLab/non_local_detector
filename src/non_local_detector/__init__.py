from non_local_detector.models import (
    ContFragClusterlessClassifier,  # noqa
    ContFragSortedSpikesClassifier,  # noqa
    MultiEnvironmentSortedSpikesClassifier,  # noqa
    MultiEnvironmentClusterlessClassifier,  # noqa
    NonLocalClusterlessDetector,  # noqa
    NonLocalSortedSpikesDetector,  # noqa
    ClusterlessDecoder,  # noqa
    SortedSpikesDecoder,  # noqa
)

try:
    from ._version import __version__
except ImportError:
    pass
