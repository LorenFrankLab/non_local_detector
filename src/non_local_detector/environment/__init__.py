from non_local_detector.environment.alignment import (
    get_2d_rotation_matrix,
    map_probabilities_to_nearest_target_bin,
)
from non_local_detector.environment.environment import Environment
from non_local_detector.environment.layout.factories import (
    get_layout_parameters,
    list_available_layouts,
)

__all__ = [
    "get_2d_rotation_matrix",
    "map_probabilities_to_nearest_target_bin",
    "Environment",
    "get_layout_parameters",
    "list_available_layouts",
]
