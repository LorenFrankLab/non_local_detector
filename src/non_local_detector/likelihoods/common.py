from dataclasses import dataclass
from typing import Optional

import jax
import jax.numpy as jnp
import numpy as np
import scipy.interpolate
from track_linearization import get_linearized_position

from non_local_detector.environment import Environment

EPS = 1e-15
LOG_EPS = np.log(EPS)


def get_position_at_time(
    time: jnp.ndarray,
    position: jnp.ndarray,
    spike_times: jnp.ndarray,
    env: Optional[Environment] = None,
):
    position_at_spike_times = scipy.interpolate.interpn(
        (time,), position, spike_times, bounds_error=False, fill_value=None
    )
    if env is not None and env.track_graph is not None:
        if position_at_spike_times.shape[0] > 0:
            position_at_spike_times = get_linearized_position(
                position_at_spike_times,
                env.track_graph,
                edge_order=env.edge_order,
                edge_spacing=env.edge_spacing,
            ).linear_position.to_numpy()[:, None]
        else:
            position_at_spike_times = jnp.array([])[:, None]

    return position_at_spike_times


@jax.jit
def gaussian_pdf(x: jnp.ndarray, mean: jnp.ndarray, sigma: jnp.ndarray) -> jnp.ndarray:
    """Compute the value of a Gaussian probability density function at x with
    given mean and sigma."""
    return jnp.exp(-0.5 * ((x - mean) / sigma) ** 2) / (sigma * jnp.sqrt(2.0 * jnp.pi))


def kde(
    eval_points: jnp.ndarray, samples: jnp.ndarray, std: jnp.ndarray
) -> jnp.ndarray:
    distance = jnp.ones((samples.shape[0], eval_points.shape[0]))

    for dim_eval_points, dim_samples, dim_std in zip(eval_points.T, samples.T, std):
        distance *= gaussian_pdf(
            jnp.expand_dims(dim_eval_points, axis=0),
            jnp.expand_dims(dim_samples, axis=1),
            dim_std,
        )
    return jnp.mean(distance, axis=0).squeeze()


def block_kde(
    eval_points: jnp.ndarray,
    samples: jnp.ndarray,
    std: jnp.ndarray,
    block_size: int = 100,
) -> jnp.ndarray:
    n_eval_points = eval_points.shape[0]
    density = jnp.zeros((n_eval_points,))
    for start_ind in range(0, n_eval_points, block_size):
        block_inds = slice(start_ind, start_ind + block_size)
        density = jax.lax.dynamic_update_slice(
            density,
            kde(eval_points[block_inds], samples, std).squeeze(),
            (start_ind,),
        )

    return density


def block_kde(
    eval_points: jnp.ndarray,
    samples: jnp.ndarray,
    std: jnp.ndarray,
    block_size: int = 100,
) -> jnp.ndarray:
    n_eval_points = eval_points.shape[0]
    density = jnp.zeros((n_eval_points,))
    for start_ind in range(0, n_eval_points, block_size):
        block_inds = slice(start_ind, start_ind + block_size)
        density = jax.lax.dynamic_update_slice(
            density,
            kde(eval_points[block_inds], samples, std).squeeze(),
            (start_ind,),
        )

    return density


@dataclass
class KDEModel:
    std: jnp.ndarray
    block_size: Optional[int] = None

    def fit(self, samples: jnp.ndarray):
        samples = jnp.asarray(samples)
        if samples.ndim == 1:
            samples = jnp.expand_dims(samples, axis=1)
        self.samples_ = samples

        return self

    def predict(self, eval_points: jnp.ndarray):
        if eval_points.ndim == 1:
            eval_points = jnp.expand_dims(eval_points, axis=1)
        std = (
            jnp.array([self.std] * eval_points.shape[1])
            if isinstance(self.std, (int, float))
            else self.std
        )
        block_size = (
            eval_points.shape[0] if self.block_size is None else self.block_size
        )

        return block_kde(eval_points, self.samples_, std, block_size)


def get_spikecount_per_time_bin(spike_times, time):
    spike_times = spike_times[
        np.logical_and(spike_times >= time[0], spike_times <= time[-1])
    ]
    return np.bincount(
        np.digitize(spike_times, time[1:-1]),
        minlength=time.shape[0],
    )
