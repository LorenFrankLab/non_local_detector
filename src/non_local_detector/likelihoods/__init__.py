from non_local_detector.likelihoods.clusterless_kde import (
    fit_clusterless_kde_encoding_model,
    predict_clusterless_kde_log_likelihood,
)
from non_local_detector.likelihoods.no_spike import (
    predict_no_spike_log_likelihood,
)  # noqa
from non_local_detector.likelihoods.sorted_spikes_glm import (
    fit_sorted_spikes_glm_encoding_model,
    predict_sorted_spikes_glm_log_likelihood,
)
from non_local_detector.likelihoods.sorted_spikes_kde import (
    fit_sorted_spikes_kde_encoding_model,
    predict_sorted_spikes_kde_log_likelihood,
)

_SORTED_SPIKES_ALGORITHMS = {
    "sorted_spikes_glm": (
        fit_sorted_spikes_glm_encoding_model,
        predict_sorted_spikes_glm_log_likelihood,
    ),
    "sorted_spikes_kde": (
        fit_sorted_spikes_kde_encoding_model,
        predict_sorted_spikes_kde_log_likelihood,
    ),
}
_CLUSTERLESS_ALGORITHMS = {
    "clusterless_kde": (
        fit_clusterless_kde_encoding_model,
        predict_clusterless_kde_log_likelihood,
    ),
}
