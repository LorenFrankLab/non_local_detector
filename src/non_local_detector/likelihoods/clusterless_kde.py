import jax
import jax.numpy as jnp
import numpy as np
from tqdm.autonotebook import tqdm
from track_linearization import get_linearized_position

from non_local_detector.environment import Environment
from non_local_detector.likelihoods.common import (
    LOG_EPS,
    EPS,
    KDEModel,
    block_kde,
    gaussian_pdf,
    get_position_at_time,
)


def get_spike_time_bin_ind(spike_times, time):
    return np.digitize(spike_times, time[1:-1])


def kde_distance(
    eval_points: jnp.ndarray, samples: jnp.ndarray, std: jnp.ndarray
) -> jnp.ndarray:
    distance = jnp.ones((samples.shape[0], eval_points.shape[0]))
    for dim_eval_points, dim_samples, dim_std in zip(eval_points.T, samples.T, std):
        distance *= gaussian_pdf(
            jnp.expand_dims(dim_eval_points, axis=0),
            jnp.expand_dims(dim_samples, axis=1),
            dim_std,
        )
    return distance


def estimate_log_joint_mark_intensity(
    decoding_spike_waveform_features: jnp.ndarray,
    encoding_spike_waveform_features: jnp.ndarray,
    waveform_stds: jnp.ndarray,
    occupancy: jnp.ndarray,
    mean_rate: float,
    position_distance: jnp.ndarray,
) -> jnp.ndarray:
    """Estimate the log joint mark intensity of decoding spikes and spike waveforms.

    Parameters
    ----------
    decoding_spike_waveform_features : jnp.ndarray, shape (n_decoding_spikes, n_features)
    encoding_spike_waveform_features : jnp.ndarray, shape (n_encoding_spikes, n_features)
    waveform_stds : jnp.ndarray, shape (n_features,)
    occupancy : jnp.ndarray, shape (n_position_bins,)
    mean_rate : float
    position_distance : jnp.ndarray, shape (n_encoding_spikes, n_position_bins)

    Returns
    -------
    log_joint_mark_intensity : jnp.ndarray, shape (n_decoding_spikes, n_position_bins)

    """
    spike_waveform_feature_distance = kde_distance(
        decoding_spike_waveform_features,
        encoding_spike_waveform_features,
        waveform_stds,
    )

    n_encoding_spikes = encoding_spike_waveform_features.shape[0]
    marginal_density = (
        spike_waveform_feature_distance.T @ position_distance / n_encoding_spikes
    )
    return jnp.log(
        mean_rate * jnp.where(occupancy > 0.0, marginal_density / occupancy, 0.0)
    )


def block_estimate_log_joint_mark_intensity(
    decoding_spike_waveform_features: jnp.ndarray,
    encoding_spike_waveform_features: jnp.ndarray,
    waveform_stds: jnp.ndarray,
    occupancy: jnp.ndarray,
    mean_rate: float,
    position_distance: jnp.ndarray,
    block_size: int = 100,
) -> jnp.ndarray:
    """Estimate the log joint mark intensity of decoding spikes and spike waveforms.

    Parameters
    ----------
    decoding_spike_waveform_features : jnp.ndarray, shape (n_decoding_spikes, n_features)
    encoding_spike_waveform_features : jnp.ndarray, shape (n_encoding_spikes, n_features)
    waveform_stds : jnp.ndarray, shape (n_features,)
    occupancy : jnp.ndarray, shape (n_position_bins,)
    mean_rate : float
    position_distance : jnp.ndarray, shape (n_encoding_spikes, n_position_bins)
    block_size : int, optional

    Returns
    -------
    log_joint_mark_intensity : jnp.ndarray, shape (n_decoding_spikes, n_position_bins)

    """
    n_decoding_spikes = decoding_spike_waveform_features.shape[0]
    n_position_bins = occupancy.shape[0]

    log_joint_mark_intensity = jnp.zeros((n_decoding_spikes, n_position_bins))

    for start_ind in range(0, n_decoding_spikes, block_size):
        block_inds = slice(start_ind, start_ind + block_size)
        log_joint_mark_intensity = jax.lax.dynamic_update_slice(
            log_joint_mark_intensity,
            estimate_log_joint_mark_intensity(
                decoding_spike_waveform_features[block_inds],
                encoding_spike_waveform_features,
                waveform_stds,
                occupancy,
                mean_rate,
                position_distance,
            ).squeeze(),
            (start_ind, 0),
        )

    return jnp.clip(log_joint_mark_intensity, a_min=LOG_EPS, a_max=None)


def fit_clusterless_kde_encoding_model(
    position_time: jnp.ndarray,
    position: jnp.ndarray,
    spike_times: list[jnp.ndarray],
    spike_waveform_features: list[jnp.ndarray],
    environment: Environment,
    sampling_frequency: int = 500,
    position_std: float = 6.0,
    waveform_std: float = 24.0,
    block_size: int = 100,
    disable_progress_bar: bool = False,
):
    position = position if position.ndim > 1 else jnp.expand_dims(position, axis=1)
    if isinstance(position_std, (int, float)):
        if environment.track_graph is not None and position.shape[1] > 1:
            position_std = jnp.array([position_std])
        else:
            position_std = jnp.array([position_std] * position.shape[1])
    if isinstance(waveform_std, (int, float)):
        waveform_std = jnp.array([waveform_std] * spike_waveform_features[0].shape[1])

    is_track_interior = environment.is_track_interior_.ravel()
    interior_place_bin_centers = environment.place_bin_centers_[is_track_interior]

    if environment.track_graph is not None and position.shape[1] > 1:
        # convert to 1D
        position1D = get_linearized_position(
            position,
            environment.track_graph,
            edge_order=environment.edge_order,
            edge_spacing=environment.edge_spacing,
        ).linear_position.to_numpy()[:, None]
        occupancy_model = KDEModel(std=position_std, block_size=block_size).fit(
            position1D
        )
    else:
        occupancy_model = KDEModel(std=position_std, block_size=block_size).fit(
            position
        )

    occupancy = occupancy_model.predict(interior_place_bin_centers)
    encoding_positions = []
    mean_rates = []
    gpi_models = []
    summed_ground_process_intensity = jnp.zeros_like(occupancy)

    n_time_bins = int((position_time[-1] - position_time[0]) * sampling_frequency)
    bounded_spike_waveform_features = []

    for electrode_spike_waveform_features, electrode_spike_times in zip(
        tqdm(
            spike_waveform_features,
            desc="Encoding models",
            unit="electrode",
            disable=disable_progress_bar,
        ),
        spike_times,
    ):
        is_in_bounds = jnp.logical_and(
            electrode_spike_times >= position_time[0],
            electrode_spike_times <= position_time[-1],
        )
        electrode_spike_times = electrode_spike_times[is_in_bounds]
        bounded_spike_waveform_features.append(
            electrode_spike_waveform_features[is_in_bounds]
        )
        mean_rates.append(len(electrode_spike_times) / n_time_bins)
        encoding_positions.append(
            get_position_at_time(
                position_time, position, electrode_spike_times, environment
            )
        )

        gpi_model = KDEModel(std=position_std, block_size=block_size).fit(
            encoding_positions[-1]
        )
        gpi_models.append(gpi_model)

        summed_ground_process_intensity += jnp.clip(
            mean_rates[-1] * gpi_model.predict(interior_place_bin_centers) / occupancy,
            a_min=EPS,
            a_max=None,
        )

    return {
        "occupancy": occupancy,
        "occupancy_model": occupancy_model,
        "gpi_models": gpi_models,
        "encoding_spike_waveform_features": bounded_spike_waveform_features,
        "encoding_positions": encoding_positions,
        "environment": environment,
        "mean_rates": mean_rates,
        "summed_ground_process_intensity": summed_ground_process_intensity,
        "position_std": position_std,
        "waveform_std": waveform_std,
        "block_size": block_size,
        "disable_progress_bar": disable_progress_bar,
    }


def predict_clusterless_kde_log_likelihood(
    time: jnp.ndarray,
    position_time: jnp.ndarray,
    position: jnp.ndarray,
    spike_times: list[jnp.ndarray],
    spike_waveform_features: list[jnp.ndarray],
    occupancy,
    occupancy_model,
    gpi_models,
    encoding_spike_waveform_features,
    encoding_positions,
    environment,
    mean_rates,
    summed_ground_process_intensity,
    position_std,
    waveform_std,
    is_local: bool = False,
    block_size: int = 100,
    disable_progress_bar: bool = False,
):
    n_time = len(time)

    if is_local:
        log_likelihood = compute_local_log_likelihood(
            time,
            position_time,
            position,
            spike_times,
            spike_waveform_features,
            occupancy_model,
            gpi_models,
            encoding_spike_waveform_features,
            encoding_positions,
            environment,
            mean_rates,
            position_std,
            waveform_std,
            block_size,
            disable_progress_bar,
        )
    else:
        is_track_interior = environment.is_track_interior_.ravel()
        interior_place_bin_centers = environment.place_bin_centers_[is_track_interior]

        log_likelihood = -1.0 * summed_ground_process_intensity * jnp.ones((n_time, 1))

        for (
            electrode_encoding_spike_waveform_features,
            electrode_encoding_positions,
            electrode_mean_rate,
            electrode_decoding_spike_waveform_features,
            electrode_spike_times,
        ) in zip(
            tqdm(
                encoding_spike_waveform_features,
                unit="electrode",
                desc="Non-Local Likelihood",
                disable=disable_progress_bar,
            ),
            encoding_positions,
            mean_rates,
            spike_waveform_features,
            spike_times,
        ):
            is_in_bounds = jnp.logical_and(
                electrode_spike_times >= time[0],
                electrode_spike_times <= time[-1],
            )
            electrode_spike_times = electrode_spike_times[is_in_bounds]
            electrode_decoding_spike_waveform_features = (
                electrode_decoding_spike_waveform_features[is_in_bounds]
            )
            position_distance = kde_distance(
                interior_place_bin_centers,
                electrode_encoding_positions,
                std=position_std,
            )
            log_likelihood += jax.ops.segment_sum(
                block_estimate_log_joint_mark_intensity(
                    electrode_decoding_spike_waveform_features,
                    electrode_encoding_spike_waveform_features,
                    waveform_std,
                    occupancy,
                    electrode_mean_rate,
                    position_distance,
                    block_size,
                ),
                get_spike_time_bin_ind(electrode_spike_times, time),
                indices_are_sorted=True,
                num_segments=n_time,
            )

    return log_likelihood


def compute_local_log_likelihood(
    time: jnp.ndarray,
    position_time: jnp.ndarray,
    position: jnp.ndarray,
    spike_times: list[jnp.ndarray],
    spike_waveform_features: list[jnp.ndarray],
    occupancy_model,
    gpi_models,
    encoding_spike_waveform_features,
    encoding_positions,
    environment,
    mean_rates,
    position_std,
    waveform_std,
    block_size: int = 100,
    disable_progress_bar: bool = False,
):
    # Need to interpolate position
    interpolated_position = get_position_at_time(
        position_time, position, time, environment
    )
    occupancy = occupancy_model.predict(interpolated_position)

    n_time = len(time)
    log_likelihood = jnp.zeros((n_time,))
    for (
        electrode_encoding_spike_waveform_features,
        electrode_encoding_positions,
        electrode_mean_rate,
        electrode_gpi_model,
        electrode_decoding_spike_waveform_features,
        electrode_spike_times,
    ) in zip(
        tqdm(
            encoding_spike_waveform_features,
            unit="electrode",
            desc="Local Likelihood",
            disable=disable_progress_bar,
        ),
        encoding_positions,
        mean_rates,
        gpi_models,
        spike_waveform_features,
        spike_times,
    ):
        is_in_bounds = jnp.logical_and(
            electrode_spike_times >= time[0],
            electrode_spike_times <= time[-1],
        )
        electrode_spike_times = electrode_spike_times[is_in_bounds]
        electrode_decoding_spike_waveform_features = (
            electrode_decoding_spike_waveform_features[is_in_bounds]
        )

        position_at_spike_time = get_position_at_time(
            position_time, position, electrode_spike_times, environment
        )

        marginal_density = block_kde(
            eval_points=jnp.concatenate(
                (
                    position_at_spike_time,
                    electrode_decoding_spike_waveform_features,
                ),
                axis=1,
            ),
            samples=jnp.concatenate(
                (
                    electrode_encoding_positions,
                    electrode_encoding_spike_waveform_features,
                ),
                axis=1,
            ),
            std=jnp.concatenate((position_std, waveform_std)),
            block_size=block_size,
        )
        occupancy_at_spike_time = occupancy_model.predict(position_at_spike_time)

        log_likelihood += jax.ops.segment_sum(
            jnp.log(
                electrode_mean_rate
                * jnp.where(
                    occupancy_at_spike_time > 0.0,
                    marginal_density / occupancy_at_spike_time,
                    0.0,
                )
            ),
            get_spike_time_bin_ind(electrode_spike_times, time),
            indices_are_sorted=True,
            num_segments=n_time,
        )

        log_likelihood -= electrode_mean_rate * jnp.where(
            occupancy > 0.0,
            electrode_gpi_model.predict(interpolated_position) / occupancy,
            0.0,
        )
    return log_likelihood[:, jnp.newaxis]
