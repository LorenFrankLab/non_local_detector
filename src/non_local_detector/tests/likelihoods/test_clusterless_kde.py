import numpy as np
import jax.numpy as jnp
import pytest

from non_local_detector.environment import Environment
from non_local_detector.likelihoods.clusterless_kde import (
    block_estimate_log_joint_mark_intensity,
    estimate_log_joint_mark_intensity,
    get_spike_time_bin_ind,
    kde_distance,
    fit_clusterless_kde_encoding_model,
    predict_clusterless_kde_log_likelihood,
)
from non_local_detector.likelihoods.common import get_position_at_time, safe_divide


def rng(seed=0):
    return np.random.default_rng(seed)


def make_simple_env_1d():
    env = Environment(environment_name="line", place_bin_size=1.0, position_range=((0.0, 10.0),))
    dummy_pos = np.linspace(0.0, 10.0, 11)[:, None]
    env = env.fit_place_grid(position=dummy_pos, infer_track_interior=False)
    assert env.place_bin_centers_ is not None
    return env


def test_get_spike_time_bin_ind_right_edge_last_bin():
    edges = np.array([0.0, 1.0, 2.0, 3.0])
    spikes = np.array([0.0, 0.5, 2.0, 3.0])
    inds = get_spike_time_bin_ind(spikes, edges)
    # bins: [0,1), [1,2), [2,3]; right-edge 3.0 -> last bin index 2
    assert inds.tolist() == [0, 0, 2, 2]


def test_kde_distance_shapes_and_values():
    r = rng(2)
    samples = r.normal(size=(7, 2))
    eval_points = r.normal(size=(5, 2))
    std = jnp.array([1.0, 0.5])
    dist = kde_distance(jnp.asarray(eval_points), jnp.asarray(samples), std)
    assert dist.shape == (samples.shape[0], eval_points.shape[0])
    assert jnp.all(dist >= 0)


def test_block_estimate_log_joint_mark_intensity_matches_unblocked():
    r = rng(3)
    dec_feats = jnp.asarray(r.normal(size=(9, 2)))
    enc_feats = jnp.asarray(r.normal(size=(15, 2)))
    weights = jnp.asarray(r.uniform(0.5, 1.5, size=(enc_feats.shape[0],)))
    wstd = jnp.array([1.0, 1.0])
    occupancy = jnp.asarray(r.uniform(0.1, 1.0, size=(6,)))
    mean_rate = 3.0
    pos_dist = jnp.asarray(r.uniform(0.0, 1.0, size=(enc_feats.shape[0], occupancy.shape[0])))

    base = estimate_log_joint_mark_intensity(
        dec_feats, enc_feats, weights, wstd, occupancy, mean_rate, pos_dist
    )
    blk = block_estimate_log_joint_mark_intensity(
        dec_feats, enc_feats, weights, wstd, occupancy, mean_rate, pos_dist, block_size=4
    )
    assert base.shape == blk.shape
    assert jnp.allclose(base, blk, rtol=1e-5, atol=1e-7)


def test_fit_and_predict_clusterless_kde_minimal():
    env = make_simple_env_1d()
    t_pos = jnp.linspace(0.0, 10.0, 101)
    pos = jnp.linspace(0.0, 10.0, 101)[:, None]
    weights = jnp.ones_like(t_pos)

    # one electrode: few encoding spikes with 2D waveform features
    enc_spike_times = jnp.array([2.0, 5.0, 7.5])
    enc_feats = jnp.array([[0.0, 0.0], [1.0, -1.0], [0.5, 0.5]], dtype=float)

    encoding = fit_clusterless_kde_encoding_model(
        position_time=t_pos,
        position=pos,
        spike_times=[enc_spike_times],
        spike_waveform_features=[enc_feats],
        environment=env,
        weights=weights,
        sampling_frequency=10,
        position_std=np.sqrt(1.0),
        waveform_std=1.0,
        block_size=8,
        disable_progress_bar=True,
    )

    for key in (
        "occupancy",
        "occupancy_model",
        "gpi_models",
        "encoding_spike_waveform_features",
        "encoding_positions",
        "encoding_spike_weights",
        "mean_rates",
        "summed_ground_process_intensity",
    ):
        assert key in encoding

    # decoding: 2 spikes within time edges with features near encoding cloud
    t_edges = jnp.linspace(0.0, 10.0, 6)
    dec_spike_times = [jnp.array([2.1, 5.2])]
    dec_feats = [jnp.array([[0.1, 0.05], [1.1, -0.9]], dtype=float)]

    ll_nonlocal = predict_clusterless_kde_log_likelihood(
        time=t_edges,
        position_time=t_pos,
        position=pos,
        spike_times=dec_spike_times,
        spike_waveform_features=dec_feats,
        occupancy=encoding["occupancy"],
        occupancy_model=encoding["occupancy_model"],
        gpi_models=encoding["gpi_models"],
        encoding_spike_waveform_features=encoding["encoding_spike_waveform_features"],
        encoding_positions=encoding["encoding_positions"],
        encoding_spike_weights=encoding["encoding_spike_weights"],
        environment=env,
        mean_rates=jnp.asarray(encoding["mean_rates"]),
        summed_ground_process_intensity=encoding["summed_ground_process_intensity"],
        position_std=jnp.asarray(encoding["position_std"]),
        waveform_std=jnp.asarray(encoding["waveform_std"]),
        is_local=False,
        block_size=8,
        disable_progress_bar=True,
    )
    # shape: (n_time, n_interior_bins)
    assert ll_nonlocal.shape[0] == t_edges.shape[0]
    assert ll_nonlocal.ndim == 2 and ll_nonlocal.shape[1] > 0
    assert jnp.all(jnp.isfinite(ll_nonlocal))

    ll_local = predict_clusterless_kde_log_likelihood(
        time=t_edges,
        position_time=t_pos,
        position=pos,
        spike_times=dec_spike_times,
        spike_waveform_features=dec_feats,
        occupancy=encoding["occupancy"],
        occupancy_model=encoding["occupancy_model"],
        gpi_models=encoding["gpi_models"],
        encoding_spike_waveform_features=encoding["encoding_spike_waveform_features"],
        encoding_positions=encoding["encoding_positions"],
        encoding_spike_weights=encoding["encoding_spike_weights"],
        environment=env,
        mean_rates=jnp.asarray(encoding["mean_rates"]),
        summed_ground_process_intensity=encoding["summed_ground_process_intensity"],
        position_std=jnp.asarray(encoding["position_std"]),
        waveform_std=jnp.asarray(encoding["waveform_std"]),
        is_local=True,
        block_size=8,
        disable_progress_bar=True,
    )
    assert ll_local.shape == (t_edges.shape[0], 1)
    assert jnp.all(jnp.isfinite(ll_local))


def test_clusterless_local_zero_spikes_equals_negative_gpi_sum():
    env = make_simple_env_1d()
    t_pos = jnp.linspace(0.0, 10.0, 101)
    pos = jnp.linspace(0.0, 10.0, 101)[:, None]
    weights = jnp.ones_like(t_pos)

    enc_spike_times = jnp.array([2.0, 5.0, 7.5])
    enc_feats = jnp.array([[0.0, 0.0], [1.0, -1.0], [0.5, 0.5]], dtype=float)

    encoding = fit_clusterless_kde_encoding_model(
        position_time=t_pos,
        position=pos,
        spike_times=[enc_spike_times],
        spike_waveform_features=[enc_feats],
        environment=env,
        weights=weights,
        sampling_frequency=10,
        position_std=np.sqrt(1.0),
        waveform_std=1.0,
        block_size=8,
        disable_progress_bar=True,
    )

    t_edges = jnp.linspace(0.0, 10.0, 6)
    empty_spikes = [jnp.array([])]
    empty_feats = [jnp.zeros((0, 2))]
    ll_local = predict_clusterless_kde_log_likelihood(
        time=t_edges,
        position_time=t_pos,
        position=pos,
        spike_times=empty_spikes,
        spike_waveform_features=empty_feats,
        occupancy=encoding["occupancy"],
        occupancy_model=encoding["occupancy_model"],
        gpi_models=encoding["gpi_models"],
        encoding_spike_waveform_features=encoding["encoding_spike_waveform_features"],
        encoding_positions=encoding["encoding_positions"],
        encoding_spike_weights=encoding["encoding_spike_weights"],
        environment=env,
        mean_rates=jnp.asarray(encoding["mean_rates"]),
        summed_ground_process_intensity=encoding["summed_ground_process_intensity"],
        position_std=jnp.asarray(encoding["position_std"]),
        waveform_std=jnp.asarray(encoding["waveform_std"]),
        is_local=True,
        block_size=8,
        disable_progress_bar=True,
    )

    interpolated_position = get_position_at_time(t_pos, pos, t_edges, env)
    occupancy_at_time = encoding["occupancy_model"].predict(interpolated_position)
    expected = -encoding["mean_rates"][0] * safe_divide(
        encoding["gpi_models"][0].predict(interpolated_position), occupancy_at_time
    )
    expected = jnp.expand_dims(expected, axis=1)
    assert jnp.allclose(ll_local, expected, rtol=1e-5, atol=1e-6)


def test_get_spike_time_bin_ind_unsorted_and_interior_edges():
    edges = np.array([0.0, 1.0, 2.0, 3.0])
    spikes = np.array([2.0, 0.0, 1.0, 1.0, 0.5])  # unsorted, includes interior edges 1.0
    inds = get_spike_time_bin_ind(spikes, edges)
    # bins: [0,1), [1,2), [2,3]; interior edge 1.0 -> bin 1 (right side)
    assert inds.tolist() == [2, 0, 1, 1, 0]


def test_fit_clusterless_kde_raises_without_place_grid():
    env = Environment(environment_name="nofit")
    t_pos = jnp.linspace(0.0, 10.0, 11)
    pos = jnp.linspace(0.0, 10.0, 11)[:, None]
    with pytest.raises(ValueError):
        fit_clusterless_kde_encoding_model(
            position_time=t_pos,
            position=pos,
            spike_times=[jnp.array([2.0])],
            spike_waveform_features=[jnp.array([[0.0, 1.0]])],
            environment=env,
        )


def test_encoding_spike_weights_and_mean_rates_match_interpolation():
    env = make_simple_env_1d()
    t_pos = jnp.linspace(0.0, 10.0, 101)
    pos = jnp.linspace(0.0, 10.0, 101)[:, None]
    # weights ramp from 0 to 1
    weights = (t_pos - t_pos.min()) / (t_pos.max() - t_pos.min())

    enc_spike_times = jnp.array([1.0, 3.0, 9.0])
    enc_feats = jnp.array([[0.0, 0.0], [1.0, -1.0], [0.5, 0.5]], dtype=float)

    encoding = fit_clusterless_kde_encoding_model(
        position_time=t_pos,
        position=pos,
        spike_times=[enc_spike_times],
        spike_waveform_features=[enc_feats],
        environment=env,
        weights=weights,
        sampling_frequency=10,
        position_std=np.sqrt(1.0),
        waveform_std=1.0,
        block_size=8,
        disable_progress_bar=True,
    )

    # Interpolated weights at spike times should match returned encoding weights
    expected_w = np.interp(enc_spike_times, t_pos, weights)
    got_w = np.asarray(encoding["encoding_spike_weights"][0])
    assert np.allclose(got_w, expected_w, rtol=1e-7, atol=1e-9)

    # Mean rate = sum(weights_at_spikes) / sum(all weights)
    expected_mean = expected_w.sum() / float(weights.sum())
    assert np.allclose(float(encoding["mean_rates"][0]), expected_mean, rtol=1e-7, atol=1e-9)
