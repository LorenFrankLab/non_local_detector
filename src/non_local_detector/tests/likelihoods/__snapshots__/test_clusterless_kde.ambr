# serializer version: 1
# name: test_clusterless_kde_encoding_model_snapshot
  dict({
    'mean_rates': list([
      0.02,
      0.025,
      0.015,
    ]),
    'n_electrodes': 3,
    'n_encoding_spikes_per_electrode': list([
      4,
      5,
      3,
    ]),
    'occupancy_stats': dict({
      'max': 0.0994294211268425,
      'mean': 0.08875985443592072,
      'min': 0.0641566589474678,
      'shape': tuple(
        10,
      ),
      'std': 0.013278350234031677,
    }),
    'summed_ground_process_intensity_stats': dict({
      'max': 0.07213348895311356,
      'mean': 0.06210869178175926,
      'min': 0.055676914751529694,
      'shape': tuple(
        10,
      ),
      'std': 0.005668018013238907,
    }),
  })
# ---
# name: test_clusterless_kde_local_likelihood_snapshot
  dict({
    'dtype': 'float32',
    'first_5': list([
      -0.005320119671523571,
      -5.756814002990723,
      -5.767607688903809,
      -0.018674537539482117,
      -0.018306149169802666,
    ]),
    'last_5': list([
      -5.756814002990723,
      -5.767607688903809,
      -0.018674537539482117,
      -0.018306149169802666,
      -0.001727289636619389,
    ]),
    'max': -0.001727289636619389,
    'mean': -1.928074836730957,
    'min': -5.767607688903809,
    'shape': tuple(
      6,
      1,
    ),
    'std': 2.7111523151397705,
    'sum': -11.568449020385742,
  })
# ---
# name: test_clusterless_kde_nonlocal_likelihood_snapshot
  dict({
    'dtype': 'float32',
    'first_5': list([
      -6.313708305358887,
      -5.946018218994141,
      -6.00192928314209,
      -6.404122829437256,
      -6.639653205871582,
    ]),
    'last_5': list([
      -0.03675157576799393,
      -0.034884270280599594,
      -0.0266812052577734,
      -0.015717117115855217,
      -0.00692210765555501,
    ]),
    'max': -0.00692210765555501,
    'mean': -4.354039669036865,
    'min': -10.636346817016602,
    'shape': tuple(
      11,
      10,
    ),
    'std': 3.3622663021087646,
    'sum': -478.9443664550781,
  })
# ---
